### PR TITLE
cannon: Support F_GETFD cmd to fcntl

### PR DESCRIPTION
**Pull request by ajsutton**
**Description**

Fixes support for go 1.22+ in cannon single threaded.

MT cannon works with go1.22 but still fails with go 1.23 still because syscall 4325.

**Tests**

Updated fuzz test expectations.

**Metadata**

- https://github.com/ethereum-optimism/optimism/issues/12049

### CODE REVIEW

**Review by clabby — APPROVED** [review]
LGTM. The manpage says the only FD flag is `FD_CLOEXEC`, which is paired up with the `execve` syscall. Cannon should never support that syscall, anyways, so should be good.

**Review by Inphi — COMMENTED** [review]
The Cannon STF change looks good. But we need to create a new `StateVersion` since we're changing VM behavior. 

I've created a PR that allows cannon to support multiple versions - https://github.com/ethereum-optimism/optimism/pull/12072. I suggest rebasing on that and updating the cannon `embeds`  so that we can continue to execute "old" cannon prior to this change.

**Comment by ajsutton** [comment]
@Inphi I can dig into this but the test failures on this are interesting.  We're dropping support for the original `singlethreaded` state but the tests are trying to dynamically generate JSON states for testing which is no longer supported.  I think the answer is to add a testutil that can generate states of specific versions and formats for us and then we can hard code examples of older states (or possibly just fudge it - we're not expecting to be able to actually read the state since its not supported, just need to detect the version).

**Comment by ajsutton** [comment]
Also I'm inclined to say that we haven't released multithreaded cannon yet so we can just add getfd support for it without a new state version.  We'll just copy in the old version for state version 0.

**Comment by Inphi** [comment]
> @Inphi I can dig into this but the test failures on this are interesting. We're dropping support for the original `singlethreaded` state but the tests are trying to dynamically generate JSON states for testing which is no longer supported. I think the answer is to add a testutil that can generate states of specific versions and formats for us and then we can hard code examples of older states (or possibly just fudge it - we're not expecting to be able to actually read the state since its not supported, just need to detect the version).

@ajsutton We probably don't need to maintain tests for older state versions. The old VM has already been tested against the v0 state. I'd port tests using state files to use the latest states rather than keep the old ones around.

**Comment by ajsutton** [comment]
> @ajsutton We probably don't need to maintain tests for older state versions. The old VM has already been tested against the v0 state. I'd port tests using state files to use the latest states rather than keep the old ones around.

@Inphi I think we need to keep the tests for op-challenger using JSON states and for detecting type 0 in Multicannon since that's important value.  But agreed we can keep it focussed on that detection which should make it much easier and all the other tests focus on the new version.

**Comment by Inphi** [comment]
Right that was the last thing I forgot to deal with. The op-challenger's using cannon packages. I need to take a look to see if we can get rid of that.

**Comment by Inphi** [comment]
So the cannon state codec needs to be backwards compatible to support this op-challenger use - https://github.com/ethereum-optimism/optimism/blob/develop/op-challenger/game/fault/trace/cannon/state_converter.go#L37.
OTOH, we have the cannon `witness` subcommand which we never use. Could the op-challenger use that instead to get the info it needs?

**Comment by ajsutton** [comment]
ugh, that's a good point actually - op-challenger is actually parsing the state.  For some of what it does it could use the witness command (and I have a patch around to do that) but it also needs to extract things like the last step.  I think `witness` can convert a state to a proof which should be enough.  I'll add it onto the yak shaving pile to get this merged... I have been quite keen to get all knowledge of cannon states out of challenger for a while so this is a forcing function.

**Comment by ajsutton** [comment]
hmm, actually I don't seem to have the final step done, but it should be implementing a `StateConverter` implementation that executes the witness subcommand.

**Comment by ajsutton** [comment]
And witness can't even generate a proof so we'll need to implement that because one of the reasons that challenger parses the state is to get the proof at the final step and it needs to actually get the step out, not just the claim in that case.

**Comment by ajsutton** [comment]
Converted to draft again until the compatibility issues are sorted. Its getting there, just a few more hairy yaks left....

**Comment by ajsutton** [comment]
I think what's left is:
* https://github.com/ethereum-optimism/optimism/pull/12137
* https://github.com/ethereum-optimism/optimism/pull/12141
* tagging a new cannon release to use to support the existing STF
* resolving all those conflicts
* updating the version detection tests to use a test utility that creates a file that looks enough like the old versions without actually having to still have the code for those old versions around.

File: ops/docker/op-stack-go/Dockerfile
Comment (Inphi) [inline ops/docker/op-stack-go/Dockerfile:54]: The `CANNON_VERSION` is still used on L59. We should preserve it.

File: packages/contracts-bedrock/src/cannon/MIPS.sol
Comment (Inphi) [inline packages/contracts-bedrock/src/cannon/MIPS.sol]: We should bump the minor version of MIPS.sol as this is a significant change.

For MIPS2.sol (MT-Cannon), it can be just a prerelease bump since it's not deployed anywhere. Though we should bump its minor version as well for consistency.

**Review by Inphi — COMMENTED** [review]
Worth considering a different naming scheme than `singlethreaded-<feature>` for new stf updates. How about `singlethreaded-<hardfork>`, so we have `singlethreaded-holocene`.
This way if we need to tweak the STF before finalizing its version, the version name is less confusing. 

This will break the `cannon-stf-job` as expected. The procedure for updating that is a bit cumbersome:
We'll need to temporarily disable that job. Merge this PR. Tag a cannon release. Then re-enable `cannon-stf-job` to verify the latest cannon release.
This rarely needs to happen, only when we create new STFs or when fixing a pre-released STF. So maybe it's no big deal. But open to suggestions on making this better.

**Comment by ajsutton** [comment]
Agreed on the naming, just not sure what naming scheme to use. These changes don't have to be tied to a hard fork since it just requires a L1 contract upgrade and not changes to the derivation rules.  I'm very tempted to just use the version number to be honest.

**Comment by Inphi** [comment]
> Agreed on the naming, just not sure what naming scheme to use. These changes don't have to be tied to a hard fork since it just requires a L1 contract upgrade and not changes to the derivation rules. I'm very tempted to just use the version number to be honest.

Version numbers would work better indeed. It's simple.

**Comment by ajsutton** [comment]
@Inphi @clabby I've gone with calling the new version `singlethreaded-2`.  I'm open to just calling all versions by just the state version number, but it felt useful to keep the singlethreaded/multithreaded information in the naming.  I think the trailing number should always be the state version number even if the `singlethreaded-*` numbers then aren't sequential.  ie we'll likely add `multithreaded-3` next and then might have `singlethreaded-4` etc.

If we do just use the StateVersion number we still need some way to name the constants in the code which is a big part of what drove me to make it `singlethreaded-2`.

File: ops/docker/op-stack-go/Dockerfile
Comment (ajsutton) [inline ops/docker/op-stack-go/Dockerfile:50]: Should we do the `cannon:v1.0.0` final release prior to merging this or stick to alpha.3 and then update after we've seen this PR merge onto develop and not get reverted for a little while?

Comment (Inphi) [reply ops/docker/op-stack-go/Dockerfile:50]: Let's stick with alpha.3

**Review by Inphi — APPROVED** [review]
LGTM!

Flagging again that this breaks the recurring [`cannon-stf-job`](https://github.com/ethereum-optimism/optimism/blob/d3402628002b7e86bf8f7e2d484e0c2f1c6ddf27/.circleci/config.yml#L1149C1-L1173C38). This is what we want, but the job will notify `@proofs-squad` when it fails. There's a chicken-and-egg problem of needing a new cannon docker image to verify `develop` against the version 2 STF. So I'm fine with leaving `cannon-stf-verify` broken until we can fix.
We can ignore the notification or disable that job for a bit. Either works.

**Comment by ajsutton** [comment]
Ah thanks I'd forgotten about that job. I might disable it as part of this PR since we'll need to get a PR up and approved to fix it and we don't need to be notifying everyone in the mean time.

**Comment by ajsutton** [comment]
hmm, looking at the config I actually don't think it will notify us - it's a dependency of something that would, but that just means the next job won't run at all and so nothing notifies.  That's worth testing to make sure we actually get notifications.  But I'll merge this tomorrow when I have more time to get a fix up (or feel free to take over in whatever way you think is best if you prefer).

**Comment by Inphi** [comment]
Merging now. I'll fix the notifications alongside adding `cannon-2` to `cannon-stf-verify`.

**Pull request merged**
